src/validators/json-schema.js

```diff
@@ -0,0 +1,318 @@
+      `The minimum value of the ${prop} must be ${validator} (current value is ${JSON.stringify(
+        val
+      )}).`,
+    maximum: (prop, val, validator) =>
+      `The maximum value of the ${prop} must be ${validator} (current value is ${JSON.stringify(
+        val
+      )}).`,
+    pattern: (prop, val, validator) =>
+      `The ${prop} value (${val}) does not match the ${validator} pattern.`,
+    maxItems: (prop, val, validator) =>
+      `The ${prop} property must not contain more than ${validator} items (currently contains ${
+        val.length
+      } items).`,
+    minItems: (prop, val, validator) =>
+      `The ${prop} property must contain at least ${validator} items (currently contains ${
+        val.length
+      } items).`,
+    divisibleBy: (prop, val, validator) =>
+      `The ${prop} property is not divisible by ${validator} (current value is ${JSON.stringify(
+        val
+      )}).`,
+    uniqueItems: (prop, val, validator) =>
+      `All items in the ${prop} property must be unique.`
+  }
+};
+
+class JsonSchema {
+  constructor(data, schema) {
+    this.data = data;
+    this.schema = schema;
+
+    if (type.string(this.data)) {
+      try {
+        this.data = JSON.parse(this.data);
+      } catch (error) {
+        outError = new errors.DataNotJsonParsableError(
+          'JSON validator: body: ' + error.message
+        );
+        outError.data = this.data;
+        throw outError;
+      }
+    }
+
+    if (type.string(this.schema)) {
+      try {
+        this.schema = JSON.parse(this.schema);
+      } catch (error) {
+        outError = new errors.SchemaNotJsonParsableError(
+          'JSON validator: schema: ' + error.message
+        );
+        outError.schema = this.schema;
+        throw outError;
+      }
+    }
+
+    this.jsonSchemaVersion = null;
+    this.validateSchema();
+  }
+
+  validateSchema() {
+    const [metaSchema, schemaVersion] = getSchemaMeta(this.schema) || [];
+
+    if (metaSchema) {
+      this.jsonSchemaVersion = schemaVersion;
+
+      if (metaSchema.$schema) {
+        tv4.reset();
+        tv4.addSchema('', metaSchema);
+        tv4.addSchema(metaSchema.$schema, metaSchema);
+        const validationResult = tv4.validateResult(this.schema, metaSchema);
+        if (!validationResult.valid) {
+          throw new errors.JsonSchemaNotValid(
+            `JSON schema is not valid draft ${this.jsonSchemaVersion}! ${
+              validationResult.error.message
+            } at path "${validationResult.error.dataPath}"`
+          );
+        }
+      }
+    } else {
+      if (metaSchemaV3.$schema) {
+        tv4.reset();
+        tv4.addSchema('', metaSchemaV3);
+        tv4.addSchema(metaSchemaV3.$schema, metaSchemaV3);
+        const validationResult = tv4.validateResult(this.schema, metaSchemaV3);
+
+        if (validationResult && validationResult.valid) {
+          this.jsonSchemaVersion = 'v3';
+          return;
+        }
+
+        tv4.reset();
```

Comment:
Outside scope of your changes, but should this reset be here and not on line 141? Invocation of this function in the case that the document is valid will cause tv4 to have some global state? Perhaps the tv4.reset at the start is protecting against this, seems this is inconsistent in code below and above and tv4.reset is always used before using tv4.

Reply:
Will remove reset here, looks inconsistent relative to other invocations. Thanks.